src/libponyc/pass/docgen.c

```diff
@@ -755,6 +798,179 @@ static void doc_methods(docgen_t* docgen, docgen_opt_t* docgen_opt,
+  // so it can be written in the mkdocs.yml file.
+  const char* doc_source_dir_relative = concat("src/", package_name);
+  doc_source_dir_relative = concat(doc_source_dir_relative, "/");
+
+  // Get relative path for [documentationDir]/src/[package_name]/[filename].md
+  const char* doc_path = concat(doc_source_dir_relative, filename_md_extension);
+
+  // Add a number (starting a 0) at the end of the file 
+  // in case there is name clashing.
+  // Only go up to 10 before not trying anymore.
+  bool are_name_clashing = false;
+  int current_numbering = 0;
+  int number_of_try = 0;
+  do {
+    are_name_clashing = false;
+    for (size_t i = 0; i < docgen->included_sources_count; i++) {
+      pony_assert(docgen->included_sources[i] != NULL);
+      // Check if the path we want to use was used for another source file.
+      if(strcmp(path, docgen->included_sources[i]->file_path) == 0) {
+        are_name_clashing = true;
+      }
+    }
+    number_of_try++;
+    // If the name are clashing add --[number] before extension to try
+    // to avoid name clashes.
+    if (are_name_clashing) {
+      current_numbering++;
+      char extension[FILENAME_MAX];
+      sprintf(extension, "--%i.md", current_numbering);
+      free((void*) filename_md_extension);
+      filename_md_extension = concat(filename_without_ext, extension);
+      path_cat(source_dir, filename_md_extension, path);
+      free((void*) doc_path);
+      doc_path = concat(doc_source_dir_relative, filename_md_extension);
+    }
+  } while (are_name_clashing && number_of_try < 10);
+
+  free((void*) filename_without_ext);
+  free((void*) filename_md_extension);
+  free((void*) doc_source_dir_relative);
+
+  if (are_name_clashing) {
+    errorf(docgen->errors, NULL, "Could not handle name clash for file %s", filename);
+    return NULL;
+  }
+
+  // Section to copy source file to [documentationDir]/src/[package_name]/[filename].md
+  FILE* file = fopen(path, "w");
+
+  if (file != NULL) {
+    // Adding a 'special' header so it can be identified 
+    // as a complete Pony source file by JavaScript  to add line numbers.
+    fprintf(file, "<div class=\"pony-full-source\" hidden> </div>\n");
+
+    // Escape markdown to tell this is Pony code
+    // Using multiple '```````'  so hopefully the markdown parser
+    // will consider the whole text as a code block.
+    fprintf(file, "```````pony\n");
+    fprintf(file, "%s", source->m);
+    fprintf(file, "\n```````");
+    fclose(file);
+
+    result->source = source;
+    result->filename = filename;
+    result->doc_path = doc_path;
+    result->file_path = path;
+    return result;
+  } else {
+    errorf(docgen->errors, NULL, "Could not write documentation to file %s", filename);
+    return NULL;
+  }
+}
+
+static char* replace_path_separator(const char* path, size_t* name_len) {
+  size_t str_len = strlen(path);
+  *name_len = str_len + 1;
+  char* buffer = (char*) ponyint_pool_alloc_size(*name_len);
```

Comment:
why using the pool allocator here and malloc/calloc for other functions? just curious.

Reply:
Well, I don't know which one I am used to use. From my C background I tend to use malloc/calloc. I should probably change them.